### PR TITLE
Make indent preserve the of \n in the input string (stop adding a trailing \n) [stable rustc]

### LINKED ISSUE #207: The indent function always adds a final \n, even if the input string has no \n
Example:

```rust
// this test fails currently
#[test]
fn indented_text_should_have_the_same_number_of_lines_as_the_original_text() {
    let texts = ["foo\nbar", "foo\nbar\n", "foo\nbar\nbaz"];
    for original in texts.iter() {
        let indented = indent(original, "");
        assert_eq!(&indented, original);
    }
}
```

There should be the same number of lines in the input and output (unless the prefix contains a newline).

[Python tests reference](https://github.com/python/cpython/blob/e42b705188271da108de42b55d9344642170aa2b/Lib/test/test_textwrap.py), [python source reference](https://github.com/python/cpython/blob/a33f2c2bae759fc9d06e1c032fd2026135f2df45/Lib/textwrap.py#L465), can't be implemented the same way in rust without the [`split_inclusive`](https://github.com/rust-lang/rust/issues/72360) feature. (cf. python's [`splitlines`](https://docs.python.org/3/library/stdtypes.html#str.splitlines))

<details>
<summary>Python script passes the same test</summary>

```python
from textwrap import indent

for original in ["foo\nbar", "foo\nbar\n", "foo\nbar\nbaz"]:
    indented = indent(original, "")
    assert original == indented
```
</details>

### PR DESCRIPTION
**Pull request by jRimbault**
I ported a custom iterator to get the inclusive split feature specifically with a `char` (though it could be even be just a custom iterator with `\n` as a const value).

This fixes #207 on stable rust but I'd feel safer once we're able to use the much more generic `split_inclusive`, hopefully.

~It uses `unsafe` code. Which is forbidden in this crate.~ It's up to you @mgeisler if you want to make an exception or not.

### CODE REVIEW

**Comment by mgeisler** [comment]
> ~It uses `unsafe` code. Which is forbidden in this crate.~ It's up to you @mgeisler if you want to make an exception or not.

Hey @jRimbault, thanks for removing the unsafe code again! Until someone shows up with a performance problem (with benchmarks...), there's no need for unsafe code here :smile:

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -12,6 +12,7 @@ env:
 jobs:
   build:
     strategy:
+      fail-fast: false
```
Comment (mgeisler) [inline .github/workflows/build.yml:15]: If we need this, then please move it to a separate PR. In general, I've been okay happy with failing PRs on the first failure since it normally indicates a real error that the PR author should fix.

File: src/indentation.rs [hunk 2]
```diff
@@ -45,20 +44,61 @@
...
 ///
-/// assert_eq!(indent(" \t  Foo   ", "->"), "-> \t  Foo   \n");
+/// assert_eq!(indent(" \t  Foo   ", "->"), "-> \t  Foo   ");
 /// ```
 pub fn indent(s: &str, prefix: &str) -> String {
     let mut result = String::new();
-    for line in s.lines() {
-        if line.chars().any(|c| !c.is_whitespace()) {
+
+    for line in CharSplitInclusive::new(s, '\n') {
```
Comment (mgeisler) [inline src/indentation.rs:52]: Could this be simplified by instead using
```rust
for idx, line in s.split('\n').enumerate() {
    // ... normal logic
    if idx > 0 {
        result.push('\n');
    }
}

I _think_ this should add exactly the name number of `\n` characters as we found originally.

File: tests/indent.rs [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+    for text in ROUNDTRIP_CASES.iter() {
+        assert_eq!(&dedent(&indent(text, " \t  \t ")), text);
+    }
+}
+
+#[test]
+fn test_indent_default() {
+    // Test default indenting of lines that are not whitespace only
+    let prefix = "  ";
+    let expected = [
```
Comment (mgeisler) [inline tests/indent.rs:59]: Is this testing something which isn't already covered in the unit tests in `src/indentations.rs`?

File: tests/indent.rs [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    for text in ROUNDTRIP_CASES.iter() {
+        assert_eq!(&indent(text, ""), text);
+    }
+    for text in WINDOWS_CASES.iter() {
+        assert_eq!(&indent(text, ""), text);
+    }
+}
+
+#[test]
+fn test_roundtrip_spaces() {
```
Comment (mgeisler) [inline tests/indent.rs:32]: Could you instead add a fuzz test for this? That would be much more powerful since it would explore many more test cases.

Please see #256 for two existing examples. You'll need [cargo-fuzz](https://github.com/rust-fuzz/cargo-fuzz) installed and then it's pretty easy to add a new fuzz test with `cargo fuzz add indent`.

File: tests/indent.rs [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+        // Pathological case
+        "  Hi.\r\n  This is a test.\n\r\n  Testing.\r\n\n",
+    ];
+    for (text, expect) in WINDOWS_CASES.iter().zip(expected.iter()) {
+        assert_eq!(&indent(text, prefix), expect)
+    }
+}
+
+#[test]
+fn indented_text_should_have_the_same_number_of_lines_as_the_original_text() {
```
Comment (mgeisler) [inline tests/indent.rs:82]: This could also be expressed very elegantly as a fuzz test.

**Comment by jRimbault** [comment]
I'll get on to it soon :)

**Comment by mgeisler** [comment]
Hi @jRimbault, no stress! That's the beauty of open source :smile:

**Comment by mgeisler** [comment]
... and thanks!

File: .github/workflows/build.yml (hunk 1 above)
Comment (jRimbault) [reply .github/workflows/build.yml:15]: Of course, this was only intended for this branch, thanks for the reminder.

File: src/indentation.rs (hunk 2 above)
Comment (jRimbault) [reply src/indentation.rs:52]: That's actually much more clever than everything I did before ^^

And it does pass the tests.

File: tests/indent.rs (hunk 3 above)
Comment (jRimbault) [reply tests/indent.rs:59]: I ported this whole test suite (file) from the tests of the python `textwrap` package, I actually don't remember the corner cases it tested, but I trust it ? :/

**Comment by jRimbault** [comment]
I tried making some fuzz work but I don't understand quite what it's testing for ? Panics ? The examples here don't seem to actually test things ?

In any case I think your solution to simply add an enumerating index is really quite good since it doesn't need bumping the MSRV (`split_inclusive` will be part of 1.51). I also think, rather than fuzzing, taking tests from tried and proved libraries like, this crate inspiration, python's `textwrap` should be a better faster option.

**Comment by mgeisler** [comment]
> I tried making some fuzz work but I don't understand quite what it's testing for ? Panics ? The examples here don't seem to actually test things ?

Yes, you raise a good point and you guessed correctly: they simply test that the code doesn't panic :smile: I should really document this...

> In any case I think your solution to simply add an enumerating index is really quite good since it doesn't need bumping the MSRV (`split_inclusive` will be part of 1.51). I also think, rather than fuzzing, taking tests from tried and proved libraries like, this crate inspiration, python's `textwrap` should be a better faster option (CPU time, and reliability wise).

Fair enough, I like it a lot that you ported those tests. They give us a nice foundation and we can then always think about turning them into property-based tests later. You're definitely right that it's much cheaper to have some solid unit tests for this.

File: tests/indent.rs (hunk 3 above)
Comment (mgeisler) [reply tests/indent.rs:59]: Okay, let's go with this then!

**Review by mgeisler — APPROVED** [review]

**Comment by jRimbault** [comment]
Happy to help 😄
